sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -1817,26 +1817,40 @@ object PushPredicateThroughNonJoin extends Rule[LogicalPlan] with PredicateHelpe
 
     case filter @ Filter(condition, union: Union) =>
       // Union could change the rows, so non-deterministic predicate can't be pushed down
-      val (pushDown, stayUp) = splitConjunctivePredicates(condition).partition(_.deterministic)
+      // We should also only push down filters which are equal (either ref or semantic) to an
+      // output of the union. We check referential equality since semantic equality of a named field
+      // may be false as the data type may have changed to include nullable during the union.
+      val output = union.output
+      def eligibleForPushdown(e: Expression): Boolean = {
```

Comment:
I'm not sure if this check is necessary. The plan is invalid if the condition references attributes that do not exist in the Union output.

Reply:
So we do need to keep the deterministic check but I _think_ you are correct we could drop the output exists check, I introduced this though as a first step to fixing the problem (did not try and push entries we could not resolve). I'll remove the check.

Reply:
nit: I think it's better to keep the code unchanged `splitConjunctivePredicates(condition).partition(_.deterministic)`, instead of creating a `eligibleForPushdown` method which does the same thing as `deterministic`.